### PR TITLE
Avoid processing `feature`s on unconfigured crates

### LINKED ISSUE #34932: In 1.11 beta, `#![cfg]` does not mask `#![feature]`
It looks like `rustc` is checking `feature` attributes *before* processing `cfg` attributes.  If so, this is a problem, because it looks like this makes it impossible to write code that optionally uses unstable features.

In the example below, the 1.10 compiler correctly ignores the test involving unstable features, but beta *does not.*  In my specific case, this means my [CI testing on beta for `newtype_derive` is unavoidably failing](https://travis-ci.org/DanielKeep/rust-custom-derive/builds/146018015).

1.12 nightly complaining about unused features is *probably* the same issue.

-----

`Cargo.toml`:

```toml
[package]
authors = ["Anonymous"]
name = "example"
version = "0.1.0"

[[test]]
name = "dummy"
path = "dummy.rs"

[features]
enable-unstable = []
```

`lib.rs`:

```rust
/* empty */
```

`dummy.rs`:

```rust
#![cfg(feature="enable-unstable")]
#![feature(iter_arith_traits)]

#[test]
fn dummy() {}
```

-----

```text
> multirust run stable rustc -vV; multirust run beta rustc -vV; multirust run nightly rustc -vV
rustc 1.10.0 (cfcb716cf 2016-07-03)
binary: rustc
commit-hash: cfcb716cf0961a7e3a4eceac828d94805cf8140b
commit-date: 2016-07-03
host: i686-pc-windows-gnu
release: 1.10.0
rustc 1.11.0-beta.1 (8dc253bcf 2016-07-05)
binary: rustc
commit-hash: 8dc253bcf3a565deb575108992b9f2e7d985aa17
commit-date: 2016-07-05
host: i686-pc-windows-gnu
release: 1.11.0-beta.1
rustc 1.12.0-nightly (7ad125c4e 2016-07-11)
binary: rustc
commit-hash: 7ad125c4eb3d620c12a868dbe77180f1a133021b
commit-date: 2016-07-11
host: i686-pc-windows-gnu
release: 1.12.0-nightly
> multirust run stable cargo test --no-default-features --features=""
   Compiling example v0.1.0 (file:///F:/Programming/Rust/sandbox/issues/beta-cfg-feature)
     Running target\debug\dummy-ae1963bd31fb7672.exe

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

> multirust run stable cargo test --no-default-features --features="enable-unstable"
   Compiling example v0.1.0 (file:///F:/Programming/Rust/sandbox/issues/beta-cfg-feature)
dummy.rs:2:1: 2:31 error: #[feature] may not be used on the stable release channel
dummy.rs:2 #![feature(iter_arith_traits)]
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
error: Could not compile `example`.

To learn more, run the command again with --verbose.
> multirust run beta cargo test --no-default-features --features=""
   Compiling example v0.1.0 (file:///F:/Programming/Rust/sandbox/issues/beta-cfg-feature)
dummy.rs:2:1: 2:31 error: #[feature] may not be used on the beta release channel [E0554]
dummy.rs:2 #![feature(iter_arith_traits)]
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
error: Could not compile `example`.

To learn more, run the command again with --verbose.
> multirust run beta cargo test --no-default-features --features="enable-unstable"
   Compiling example v0.1.0 (file:///F:/Programming/Rust/sandbox/issues/beta-cfg-feature)
dummy.rs:2:1: 2:31 error: #[feature] may not be used on the beta release channel [E0554]
dummy.rs:2 #![feature(iter_arith_traits)]
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
error: Could not compile `example`.

To learn more, run the command again with --verbose.
> multirust run nightly cargo test --no-default-features --features=""
   Compiling example v0.1.0 (file:///F:/Programming/Rust/sandbox/issues/beta-cfg-feature)
dummy.rs:2:12: 2:29 warning: unused or unknown feature, #[warn(unused_features)] on by default
dummy.rs:2 #![feature(iter_arith_traits)]
                      ^~~~~~~~~~~~~~~~~
     Running target\debug\dummy-ae1963bd31fb7672.exe

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

> multirust run nightly cargo test --no-default-features --features="enable-unstable"
   Compiling example v0.1.0 (file:///F:/Programming/Rust/sandbox/issues/beta-cfg-feature)
dummy.rs:2:12: 2:29 warning: unused or unknown feature, #[warn(unused_features)] on by default
dummy.rs:2 #![feature(iter_arith_traits)]
                      ^~~~~~~~~~~~~~~~~
     Running target\debug\dummy-ae1963bd31fb7672.exe

running 1 test
test dummy ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
```

### PR DESCRIPTION
**Pull request by jseyfried**
Fixes #34932, a regression caused by #34272.
r? @nrc

### CODE REVIEW

**Comment by retep998** [comment]
Why a compile fail test? Isn't the idea that disabling the crate with `#![cfg]` should allow it to compile despite use of `#![feature]`?

**Comment by jseyfried** [comment]
@retep998 `feature` doesn't prevent compilation on the nightlies since it is only disabled on stable/beta.

This test still tests that `feature` on an unconfigured crate is not processed. On stable/beta, processing `feature` causes an error (since `feature` is disabled), but on nightlies, it can *prevents* gated feature errors.

**Comment by retep998** [comment]
For example
```
#![cfg(something)]
#![feature(some_feature)]
```
If `something` isn't set, it will compile on the current stable. However it will _not_ compile in the current beta. The regression is that the stability check for features is done even when the entire crate is disabled with a cfg.

So the correct test would need to make sure that example would still compile in a stable environment where features are disabled.

**Comment by jseyfried** [comment]
@retep998 I know, this fixes that regression. The test I added is equivalent since features are stability checked when they are processed, so testing that features are not processed also tests that they are not stability checked.

If you know of a more direct way to test this, let me know -- I'm not sure how to add a test with stable/beta feature stability checking.

**Comment by eddyb** [comment]
```
// compile-flags:--test
// rustc-env:RUSTC_BOOTSTRAP_KEY=

#![cfg(feature="does-not-exist")]
#![feature(iter_arith_traits)]

// nothing below actually matters, but I added it anyway
#[test]
fn dummy() {
    let () = "this should not reach type-checking";
    panic!("this should not run");
}
```
At least when building beta/stable, this test will require no `feature`'s to be observed, in order to compile.
Although the `feature` being used should be something we know we aren't going to ever stabilize (`rustc_attrs`?).

**Comment by retep998** [comment]
You could add a `#![deny(unused_features)]` to make sure it passes on nightly as well, since currently on nightly it passes with a warning about an unused feature.

**Comment by jseyfried** [comment]
@eddyb @retep998 I amended the regression test with your suggestions.

**Comment by eddyb** [comment]
LGTM. cc @alexcrichton Is this the best we can do about testing?

**Comment by alexcrichton** [comment]
I don't have too many opinions, so for me if it works it works!

File: src/test/run-pass/issue-34932.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags:--test
+// rustc-env:RUSTC_BOOTSTRAP_KEY=
```
Comment (eddyb) [inline src/test/run-pass/issue-34932.rs]: You'll also need (found by Travis):
```rust
// ignore-pretty : (#23623) problems when  ending with // comments
```

Comment (jseyfried) [inline src/test/run-pass/issue-34932.rs]: Done.

**Comment by retep998** [comment]
What, no information from @bors about how the build went?

**Comment by nikomatsakis** [comment]
Accepting for backport: regression, small.

**Pull request merged**